### PR TITLE
enable the use of an apiUse block in an apiDefine block 

### PR DESCRIPTION
**Pull request by tommy87**
Enabled the use of an apiUse block in an apiDefine block

see issue https://github.com/apidoc/apidoc/issues/859

Note:
I have moved my changes to my dev branch and create a new pull request to apply only my changes. The old request was https://github.com/apidoc/apidoc-core/pull/96

### CODE REVIEW

File: lib/workers/api_use.js [hunk 1]
```diff
@@ -66,70 +66,86 @@ function postProcess(parsedFiles, filenames, preProcess, packageInfos, source, t
...
+	        }
+                
+                //create a copy of the elements for save iterating of the elements
+                const blockClone = [...block.local[target]];
 
-            block.local[target].forEach(function(definition) {
-                var name = definition.name;
-                var version = block.version || packageInfos.defaultVersion;
+                // remove unneeded target before starting the loop, to allow a save insertion of new elements
+                // TODO: create a cleanup filter
```
Comment (NicolasCARPi) [inline lib/workers/api_use.js:88]: could you expand on that? It's not great to add a TODO ;)

Comment (tommy87) [reply lib/workers/api_use.js:88]: This is not from me. I only moved it to another place

Comment (tommy87) [reply lib/workers/api_use.js:88]: In the connected ticket you can see the changes i made in a better overview. Githubs pullreuest didnt resolve the extra space i add for the new while loop

File: lib/parser.js [hunk 2]
```diff
@@ -222,7 +222,8 @@ Parser.prototype._parseBlockElements = function(indexApiBlocks, detectedElements
                     preventGlobal = elementParser.preventGlobal === true;
 
                     // allow multiple inserts into pathTo
-                    allowMultiple = elementParser.allowMultiple === true;
+                    //allowMultiple = elementParser.allowMultiple === true;
```
Comment (NicolasCARPi) [inline lib/parser.js:225]: no need to keep the old code commented ;)

File: lib/workers/api_use.js [hunk 3]
```diff
@@ -66,70 +66,86 @@ function postProcess(parsedFiles, filenames, preProcess, packageInfos, source, t
 
     parsedFiles.forEach(function(parsedFile, parsedFileIndex) {
         parsedFile.forEach(function(block) {
-            if ( ! block.local[target])
-                return;
+            var loopCounter = 0;     //add a loop counter to have a break condition when the recursion depth exceed a predifined limit
+	    while (block.local[target]) {
```
Comment (NicolasCARPi) [inline lib/workers/api_use.js:70]: use 4 spaces indent, no tabs

File: lib/parser.js (hunk 2 above)
Comment (tommy87) [reply lib/parser.js]: i have removed it now

File: lib/workers/api_use.js (hunk 3 above)
Comment (tommy87) [reply lib/workers/api_use.js]: i replaced all tabs by 4 spaces indent

**Comment by tommy87** [comment]
@NicolasCARPi 
is there anything else i need todo? If not it would be great if you could merge my commits

**Comment by NicolasCARPi** [comment]
I'm sorry if it takes a bit of time for PRs to be merged but I need to test them first. Speaking of tests, it would be great if you could add some test for that new feature in the test suite!

**Comment by tommy87** [comment]
I can add some tests, this shoulnd be a problem. But you need to tell me where i can add the test.

**Comment by NicolasCARPi** [comment]
In the `test` folder ;)

**Comment by tommy87** [comment]
@NicolasCARPi 
Okay i have now fixed errors shown by jshint and also add two test cases for the apiUse.

I think it would also be an big improvement for the project, if you can enable continous integration, which just run the "test" script

File: test/worker_api_use_test.js [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    });
+
+    it('case 2: recursive test', function(done) {
+        var preProcess = worker.preProcess(parsedFilesRecursiveTest, filenames, packageInfos);
+        worker.postProcess(parsedFilesRecursiveTest, filenames, preProcess, packageInfos);
+        
+        var rootBlock = parsedFilesRecursiveTest[0][0];
+        rootBlock.local.name.should.eql('root');
+
+        //console.log(rootBlock);
```
Comment (NicolasCARPi) [inline test/worker_api_use_test.js:202]: this should be removed

**Comment by tommy87** [comment]
@NicolasCARPi still waiting for the merge

If you need help with this project, i could also help as a reviewer/maintainer. At least we want to use this tool more often in our company and i really waiting for the current merge, do enable auto generation of the documentation in our CI

File: lib/workers/api_use.js [hunk 5]
```diff
@@ -66,70 +66,88 @@ function postProcess(parsedFiles, filenames, preProcess, packageInfos, source, t
...
                         messages.common.element,
                         messages.common.usage,
                         messages.common.example,
                         [
-                            { 'Groupname': name }
+                           { 'Groupname': block.name }
                         ]
                     );
                 }
+
```
Comment (NicolasCARPi) [inline lib/workers/api_use.js:83]: this line has whitespaces (we should really have an automatic tool to look/fix these minor issues...)

**Review by NicolasCARPi — APPROVED** [review]

**Comment by NicolasCARPi** [comment]
Hello @tommy87, you did well to remind me. Your code has been merged (sorry for the delay). I'll publish a new version soon. Ping me if not.

**Comment by NicolasCARPi** [comment]
> If you need help with this project, i could also help as a reviewer/maintainer.

Your help would be very welcome. See with @rottmann if he's willing to give you some permissions. Even without that, please feel free to review existing PRs/issues and comment on them.

**Comment by tommy87** [comment]
> Hello @tommy87, you did well to remind me. Your code has been merged (sorry for the delay). I'll publish a new version soon. Ping me if not.

Ping ;)

**Comment by tommy87** [comment]
@rottmann if you want you can add me to the list of reviewers. So i can help you a little bit

**Comment by rottmann** [comment]
@tommy Welcome on board, i give you for the start "triage" rights, so you can manage issues and pull requests. Please coordinage with Nicolas, he is the "head of the department". 

@NicolasCARPi Upraded you rights to maintainer. So you can manage more settings in repositories.

**Comment by NicolasCARPi** [comment]
@rottmann Noted!

@tommy87 Welcome! Please look at the open PRs, test the code and report if you think it's good and can be merged or if you see any issue!

**Pull request merged**